### PR TITLE
Add "file sources" to Iodide

### PR DESCRIPTION
**Pull request by wlach**
### Pull Request checklist

- [x] **Documentation**: If this feature has or requires documentation, the relevant docs have been updated.
- [x] **Changelog**: This PR updates the [changelog](../CHANGELOG.md) with any user-visible changes.
- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not

### CODE REVIEW

**Comment by wlach** [comment]
@jezdez this isn't nearly complete, but I'd love your opinion on the celery/backend work on this to-date if you have a spare moment. Context is issue #1976

File: server/notebooks/views.py [hunk 1]
```diff
@@ -29,7 +29,9 @@ def _get_iframe_src():
 
 @ensure_csrf_cookie
 def notebook_view(request, pk):
+
     notebook = get_object_or_404(Notebook, pk=pk)
+
```
Comment (wlach) [inline server/notebooks/views.py]: Looks like we added some newlines to this file by mistake.

File: server/tests/test_file_source_api.py [hunk 2]
```diff
@@ -0,0 +1,128 @@
+from datetime import timedelta
+
+import pytest
+from django.urls import reverse
+
+from server.files.models import FileSource
+
+
```
Comment (wlach) [inline server/tests/test_file_source_api.py]: We should add some tests for listing file sources here...

File: server/tests/test_notebook_view.py [hunk 3]
```diff
@@ -56,11 +57,11 @@ def test_notebook_view_old_revision(client, test_notebook):
     )
     assert resp.status_code == 200
     assert get_title_block(resp.content) == initial_revision.title
-    print(str(resp.content))
     assert get_script_block(resp.content, "iomd", "text/iomd") == initial_revision.content
     assert get_script_block_json(resp.content, "notebookInfo") == {
         "connectionMode": "SERVER",
         "files": [],
+        "fileSources": [],
```
Comment (wlach) [inline server/tests/test_notebook_view.py]: I think this might be referencing something from the notebook view which is no longer there?

In any case we should probably settle on using snake case for these properties, just to be consistent with what's there already (we could talk about modifying both these views and the API to use camel case for better integration on the client side, but that's probably best settled elsewhere).

Comment (hamilton) [reply server/tests/test_notebook_view.py]: yes this should be removed actually. We don't ship `fileSources` to the notebook in the notebook view anymore.

File: server/tests/test_file_source_api.py (hunk 2 above)
Comment (hamilton) [reply server/tests/test_file_source_api.py]: +1

File: server/tests/test_file_source_api.py [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+        reverse("notebook-file-sources-list", kwargs={"notebook_id": test_notebook.id})
+    )
+    assert resp.status_code == 200
+    assert resp.json() == [
+        {
+            "filename": test_file_source.filename,
+            "id": test_file_source.id,
+            "latest_file_update_operation": None,
+            "update_interval": str(test_file_source.update_interval.total_seconds()),
+        }
```
Comment (wlach) [inline server/tests/test_file_source_api.py]: This test is currently failing, because we need to specify a different serializer in case the user is not the owner of the notebook, which does not show the url

File: server/files/serializers.py [hunk 5]
```diff
@@ -1,9 +1,85 @@
...
+    """
+    Gives a summary of file (not its content)
+    """
+
     class Meta:
         model = File
         fields = ("id", "notebook_id", "filename", "last_updated")
+
+
+class FileSourceSerializer(serializers.ModelSerializer):
```
Comment (wlach) [inline server/files/serializers.py]: Should have a version of this serializer for the not-logged-in case, without the url field

File: server/files/api_views.py [hunk 6]
```diff
@@ -53,3 +55,71 @@ def update(self, request, pk):
         return Response(FilesSerializer(file_obj_to_update).data, status=201)
 
     queryset = File.objects.all()
+
+
+class NotebookFileSourceViewSet(viewsets.ModelViewSet):
+
+    serializer_class = FileSourceDetailSerializer
+    http_method_names = ["get"]
+
```
Comment (wlach) [inline server/files/api_views.py]: We should be able to use the `get_serializer_class` method here to pick the right serializer, depending on whether the user is logged in or not.

File: server/notebooks/views.py (hunk 1 above)
Comment (hamilton) [reply server/notebooks/views.py]: hmm, there was a story of another line of code here once upon a time, but that story is lost to the world. alas

File: src/editor/components/modals/file-modal/manage-file-sources/add-new-file-source.jsx [hunk 7]
```diff
@@ -0,0 +1,225 @@
...
+  color: ${props => (props.statusType === "ERROR" ? "red" : "black")};
+`;
+
+const UpdateIntervalContainer = styled.div`
+  span {
+    font-size: 13px;
+  }
+`;
+
+const UpdateIntervalSelector = styled.select`
```
Comment (bcolloran) [inline src/editor/components/modals/file-modal/manage-file-sources/add-new-file-source.jsx]: the styling and behavior of this button don't really match the other buttons in the rest of the app (using a system default `select` menu rather than a custom drop down, the corner radius is different, the text doesn't match the all caps that we have in other buttons... @hamilton, i wonder if we can reuse some of the custom components you worked on earlier this year rather than this one-off code for `UpdateIntervalContainer` (and the stuff supporting it). 

I also notice that the code here with the background image svg seems rather one-off and hard to follow compared to using a MUI icon in a component. i bet we can clean this out with minimal muss and fuss

 (as well as the code here with the background image svg)

File: src/editor/components/modals/file-modal/manage-file-sources/file-source-list.jsx [hunk 8]
```diff
@@ -0,0 +1,278 @@
...
+
+import {
+  List,
+  ListItem,
+  ListMain,
+  ListPrimaryText,
+  ListSecondaryText,
+  ListSecondaryTextLink,
+  ListMetadata
+} from "../../../../../server/components/list";
```
Comment (bcolloran) [inline src/editor/components/modals/file-modal/manage-file-sources/file-source-list.jsx]: should refactor all this to be in `shared`?

File: src/editor/components/modals/file-modal/manage-file-sources/file-source-list.jsx [hunk 9]
```diff
@@ -0,0 +1,278 @@
+import React, { useState, useEffect } from "react";
+import styled from "react-emotion";
+import PropTypes from "prop-types";
+import { connect } from "react-redux";
+
+import format from "date-fns/format";
+import CircularProgress from "@material-ui/core/CircularProgress";
+
+import DeleteModal from "../../../../../server/components/delete-modal";
```
Comment (bcolloran) [inline src/editor/components/modals/file-modal/manage-file-sources/file-source-list.jsx]: move to `shared`?

File: src/editor/components/modals/file-modal/manage-file-sources/file-source-list.jsx [hunk 10]
```diff
@@ -0,0 +1,278 @@
...
+  position: relative;
+  height: 100%;
+  display: grid;
+  margin: auto;
+  align-content: center;
+  overflow: hidden;
+  cursor: pointer;
+`;
+
+const Fader = styled.div`
```
Comment (bcolloran) [inline src/editor/components/modals/file-modal/manage-file-sources/file-source-list.jsx]: i think it would be worth breaking something like fader into a reusable component in our component library, something like "SpinnerButton" or whatever you want to call it. it's a nice little thing for giving a bit of feedback, but it seems like a generic thing that doesn't belong in this file.

but instead of inventing something from scratch, perhaps we should build a "SpinnerButton" component for our library based on the "Interactive Integration" button from MUI here: https://material-ui.com/components/progress/ (but without the circle icon button, just the text version). it gives the same feedback without more custom code.

**Comment by teonbrooks** [comment]
cc: @jtg567

File: src/editor/components/modals/file-modal/manage-file-sources/add-new-file-source.jsx [hunk 11]
```diff
@@ -0,0 +1,225 @@
...
+  );
+}
+
+addNewFileSourceUnconnected.propTypes = {
+  addNewFileSource: PropTypes.func
+};
+
+function mapStateToProps(state) {
+  return {
+    filesSources: state.notebookInfo.filesSources
```
[the remark this answers is not in the document]
Comment (hamilton) [reply src/editor/components/modals/file-modal/manage-file-sources/add-new-file-source.jsx]: good catch

File: src/editor/components/modals/file-modal/manage-file-sources/file-source-list.jsx [hunk 12]
```diff
@@ -0,0 +1,278 @@
...
+                <ListMain>
+                  <ListPrimaryText>
+                    <FileInformationContainer>
+                      <FileName
+                        hasBeenRun={fileSource.latest_file_update_operation}
+                      >
+                        {fileSource.filename}
+                      </FileName>
+                      {fileSource.latest_file_update_operation ? (
+                        <StatusLabel
```
[the remark this answers is not in the document]
Comment (hamilton) [reply src/editor/components/modals/file-modal/manage-file-sources/file-source-list.jsx]: good idea

File: src/editor/state-schemas/state-schema.js [hunk 13]
```diff
@@ -13,6 +13,33 @@ export const FETCH_CHUNK_TYPES = [
...
+export const FILE_SOURCE_UPDATE_INTERVALS = Object.keys(
+  FILE_SOURCE_UPDATE_INTERVAL_MAP
+);
+export const FILE_UPDATE_OPERATION_STATUSES = [
+  "pending",
+  "running",
+  "completed",
+  "failed"
+];
+
```
[the remark this answers is not in the document]
Comment (hamilton) [reply src/editor/state-schemas/state-schema.js]: will do

File: src/editor/state-schemas/state-schema.js [hunk 14]
```diff
@@ -51,6 +78,29 @@ export const fileSchema = {
...
+};
+
+export const fileSourceSchema = {
+  type: "object",
+  properties: {
+    id: { type: "integer" },
+    url: { type: ["string", "null"] },
+    filename: { type: "string" },
+    update_interval: { type: ["string", "null"] },
+    last_file_update_operation: fileUpdateOperationSchema
```
[the remark this answers is not in the document]
Comment (hamilton) [reply src/editor/state-schemas/state-schema.js]: this is used in the component, not sure what you mean.

File: src/editor/reducers/file-source-reducer.js [hunk 15]
```diff
@@ -0,0 +1,84 @@
...
+};
+
+export default (state, action) => {
+  // FIXME: remove reduceReducers in reducer.js and have state only reflect
+  // the app's fileSources object. This will require refactoring createValidatedReducer
+  // but should be considered a high priority.
+  switch (action.type) {
+    case "UPDATE_FILE_SOURCES": {
+      if (!Array.isArray(action.fileSources))
+        throw Error(
```
[the remark this answers is not in the document]
Comment (hamilton) [reply src/editor/reducers/file-source-reducer.js]: Throwing is a side effect?

File: src/editor/components/modals/file-modal/manage-file-sources/file-source-list.jsx (hunk 8 above)
Comment (hamilton) [reply src/editor/components/modals/file-modal/manage-file-sources/file-source-list.jsx]: yeah, probably time to do that

File: src/editor/reducers/file-source-reducer.js (hunk 15 above)
[the remark this answers is not in the document]
Comment (hamilton) [reply src/editor/reducers/file-source-reducer.js]: hmm, rather than wade into that tarpit, I'll just remove this :)

[the remark this answers is not in the document]
Comment (bcolloran) [reply src/editor/reducers/file-source-reducer.js]: yep :-)

throwing causes an effect that is visible outside the function. or, to put it another way: it effects the world in some way other than just through the value that the function returns.

in this case, i think throwing would also prevent the reducer from returning a value, in which case there would be no next state. i'm not sure, would that effectively halt the app? i think we would want to intercept something like that upstream (in a thunk or whatever). or if we wanted the user to know, we could change the state based on that condition

File: src/editor/components/modals/file-modal/manage-files/index.jsx [hunk 16]
```diff
@@ -0,0 +1,352 @@
...
+  static defaultProps = {
+    maxFileSize: false,
+    maxFileSizeMB: false,
+    maxFilenameLength: false,
+    savedFiles: []
+  };
+
+  constructor(props) {
+    super(props);
+    this.state = {
```
[the remark this answers is not in the document]
Comment (hamilton) [reply src/editor/components/modals/file-modal/manage-files/index.jsx]: This file was merely moved (so not directly part of this PR), so this can be addressed when we touch `manage-files` again (which will be soon)

File: src/editor/components/modals/file-modal/manage-files/index.jsx [hunk 17]
```diff
@@ -0,0 +1,352 @@
...
+        onAddButtonClick={this.onAddButtonClick}
+        confirmDelete={this.confirmDelete}
+      />
+    </React.Fragment>
+  );
+}
+
+export function mapDispatchToProps(dispatch) {
+  return {
+    saveFile: async (formData, fileID = undefined) => {
```
[the remark this answers is not in the document]
Comment (hamilton) [reply src/editor/components/modals/file-modal/manage-files/index.jsx]: this file as moved, so we can address this when we rewrite the `manage-files` part

File: src/editor/components/modals/file-modal/manage-file-sources/index.jsx [hunk 18]
```diff
@@ -0,0 +1,21 @@
+import React from "react";
+import styled from "react-emotion";
+import AddNewFileSource from "./add-new-file-source";
+import FileSourceList from "./file-source-list";
+
+const FileSourceContainer = styled.div`
+  --marg: 20px;
+  padding: var(--marg);
```
[the remark this answers is not in the document]
Comment (hamilton) [reply src/editor/components/modals/file-modal/manage-file-sources/index.jsx]: This variable is used in all the children of this component as well

**Comment by hamilton** [comment]
@wlach seems like we should do some filename validation on the server side rather than having the client do that (cc @bcolloran). Not sure what that sound look like

File: src/editor/components/modals/file-modal/manage-file-sources/add-new-file-source.jsx (hunk 7 above)
Comment (hamilton) [reply src/editor/components/modals/file-modal/manage-file-sources/add-new-file-source.jsx]: yep will do

**Comment by hamilton** [comment]
Thanks for the review @bcolloran.

- docs: @wlach and I discussed whether to have docs here, and of course we will, but it has become apparent that the UX here is going to change once we merge files with file sources, so we decided not to for now. Happy to change that opinion.
- state: we discussed the other components over slack, but one thing I do want to stress – the `manage-files` component was merely moved, not added to this PR (as you can see from `master`). Given the above we chose not to refactor that component here. That feature has a bunch of local state that will disappear upon the refactor.
- "RUN NOW" and "DELETE" are both buttons, following the standard material design practice of stressing the important one with either outline or contained button, and de-emphasizing the secondary button. Happy to change that if you have a suggestion.
- filename validation: I think this should probably be done on the server and have that error be pushed to the UI here. @wlach do you have thoughts about how to answer the question "what is a valid filename"?

As for your other comments, I'll tweak or I've left comments accordingly. There are a lot of little choices that were made intentionally or are a new thing we need to figure out (for instance, the server API returns snake case directly and I'm not sure how to transform it to camel case from Django atm, the store should probably be consistent there). But we can leave that for a second review or perhaps an irl conversation, since I think GH PR reviews don't really leave much room for meaningful discussion

File: src/editor/components/modals/file-modal/manage-file-sources/file-source-list.jsx (hunk 10 above)
Comment (hamilton) [reply src/editor/components/modals/file-modal/manage-file-sources/file-source-list.jsx]: This is pretty much exactly what that example you linked to already does, I think

Comment (hamilton) [reply src/editor/components/modals/file-modal/manage-file-sources/file-source-list.jsx]: happy to separate this out as another component though

File: src/editor/components/modals/file-modal/manage-file-sources/file-source-list.jsx [hunk 19]
```diff
@@ -0,0 +1,278 @@
...
+`;
+
+const LastRanLabel = styled.div`
+  padding-top: 2px;
+  padding-bottom: 2px;
+  font-size: 12px;
+  color: gray;
+`;
+
+const operationStatusColors = {
```
[the remark this answers is not in the document]
Comment (hamilton) [reply src/editor/components/modals/file-modal/manage-file-sources/file-source-list.jsx]: Not a bad idea, but I'd prefer to not do that quite yet until there is a more compelling reason, I think. For instance the colors in the kernel spinner are against a dark background, so they don't 1:1 map perfectly to these

File: src/editor/state-schemas/state-schema.js [hunk 20]
```diff
@@ -51,6 +78,29 @@ export const fileSchema = {
...
 
+export const fileUpdateOperationSchema = {
+  type: "object",
+  properties: {
+    id: { type: "integer" },
+    scheduled: { type: "string" },
+    started: { type: "string" },
+    ended: { type: "string" },
+    status: { type: "integer" },
+    failure_reason: { type: "string" }
```
[the remark this answers is not in the document]
Comment (hamilton) [reply src/editor/state-schemas/state-schema.js]: `failure_reason` will be added to the file source list component to satisfy one of the other pieces of feedback you've given (and was on the to-do list for this PR)

File: src/editor/state-schemas/state-schema.js (hunk 14 above)
[the remark this answers is not in the document]
Comment (hamilton) [reply src/editor/state-schemas/state-schema.js]: We aren't using local hooks state management for any part of the `fileSources` stuff other than UX state

**Comment by hamilton** [comment]
hmmm, looks like there are a _lot_ of problems with the source branch so I'm going to need to chat w/ will about this more tomorrow. I can't seem to pull his changes in without resolving a lot of merge conflicts.

**Comment by hamilton** [comment]
oops – pushing to the wrong branch. Need to fix all the merge conflicts from wlach's progress first :)

**Comment by wlach** [comment]
@hamilton I rebased the branch just before I left last week and squashed most of the commits. 

It looks like you're still working against the old version with a million commits. I did warn you that I was going to do the squashing/rebasing, but that admittedly was a while ago. :)

We should probably just cherry pick your last commits onto the rebased/squashed branch (which is also based on a more recent version of master). 

Will look into the other issues tomorrow

**Comment by hamilton** [comment]
@wlach as you can see I fixed it :) I thought I had pulled before getting to work, so imagine my surprise when I pushed to your branch.

**Comment by hamilton** [comment]
hmm, we'll figure that out tomorrow. Thought I had pulled before starting, but perhaps not. I fixed those merge conflicts and then pushed, nothing that can't be fixed though.

**Comment by wlach** [comment]
As for filename validation, we should probably do it on both the client and the server, there are very few rules about what filenames are allowed in unix land, but perhaps we should err on the side of being restrictive. This stack overflow article has some decent suggestions:

https://stackoverflow.com/a/458001

I don't know if this functionality needs to be in the initial landing, or we can do a followup

**Comment by hamilton** [comment]
@bcolloran the frontend part of this PR is ready for another pair of eyes. This feature was in hindsight unnecessarily laborious to implement / fix up, and I'm realizing it's really because we don't have a good process in-place for this kind of UX work. I have some ideas on how to make this process easier the next time around based on this feature. For the time being, I did my best to take care of as much as I could given the current set of constraints.

File: docs/workflows.md [hunk 21]
```diff
@@ -43,11 +45,37 @@ syntax --
...
-
-This notebook [shows how to load and use an image blob](https://alpha.iodide.io/notebooks/127/).
+-- where `TYPE` is the type of the file ("text", "json", "arraybuffer", or
+"blob"), `VARNAME` is the variable into which you want to load the data, and
+`FILE_NAME` is the name of the file you uploaded.
+
+This notebook [shows how to load and use an image
+blob](https://alpha.iodide.io/notebooks/127/).
+
+### scheduling file uploads
```
Comment (hamilton) [inline docs/workflows.md]: On request, here is some documentation. Feel free to insert suggestions rather than comment on this to save time.

File: src/editor/components/modals/file-modal/manage-file-sources/__stories__/manage-file-sources.stories.js [hunk 22]
```diff
@@ -0,0 +1,136 @@
+import React, { useState } from "react";
```
Comment (hamilton) [inline src/editor/components/modals/file-modal/manage-file-sources/__stories__/manage-file-sources.stories.js]: I realized only a couple of hours ago that we need to be using Storybooks for _all_ component development, including this one. We don't have an accepted way of organizing these storybooks, so read through this file with that caveat in mind. I was not comprehensive by _any_ means with this one – merely playing with the idea and adding the components that I needed to really decouple from the app for further development to get this more cleaned up. A more thorough PR would incorporate the other parts of the `ManageFileSources` component as well as the component _in toto_. I'm happy to do that down the line but probably not for this PR.

File: src/editor/components/modals/file-modal/manage-file-sources/__tests__/file-source-list-item.test.js [hunk 23]
```diff
@@ -0,0 +1,135 @@
...
+  // hasn't run
+  filename: "hasntRunYet",
+  id: 0,
+  url: "https://whatever.com/api",
+  update_interval: "never",
+  latest_file_update_operation: {}
+};
+
+const currentlyRunning = {
+  // hasn't run
```
Comment (hamilton) [inline src/editor/components/modals/file-modal/manage-file-sources/__tests__/file-source-list-item.test.js]:
```suggestion
```

File: src/editor/components/modals/file-modal/manage-file-sources/add-new-file-source.jsx [hunk 24]
```diff
@@ -0,0 +1,202 @@
...
+
+const ALLOWED_PROTOCOLS = ["https", "http"];
+
+const hasAllowedProtocol = url => {
+  return ALLOWED_PROTOCOLS.some(protocol => {
+    return url.startsWith(protocol);
+  });
+};
+
+export function AddNewFileSourceUnconnected({
```
Comment (hamilton) [inline src/editor/components/modals/file-modal/manage-file-sources/add-new-file-source.jsx]: note: this component still does not perform validation of the URL nor the filename.

File: src/editor/components/modals/file-modal/manage-file-sources/add-new-file-source.jsx [hunk 25]
```diff
@@ -0,0 +1,202 @@
...
+  addFileSource
+}) {
+  const [statusVisible, updateStatusVisibility] = useState(false);
+  const [status, updateStatus] = useState({ type: "NONE", text: "" });
+
+  const handleUpdateIntervalChange = event => {
+    updateFileSourceInputUpdateInterval(event.target.value);
+  };
+  const submitInformation = async () => {
+    updateStatusVisibility(true);
```
Comment (hamilton) [inline src/editor/components/modals/file-modal/manage-file-sources/add-new-file-source.jsx]: on further inspection, since it was requested we put all the state in this component in redux, I should probably move the status reporting there as well.

File: src/editor/components/modals/file-modal/manage-file-sources/in-progress.jsx [hunk 26]
```diff
@@ -0,0 +1,36 @@
+import React from "react";
```
Comment (hamilton) [inline src/editor/components/modals/file-modal/manage-file-sources/in-progress.jsx]: I think with a more thorough Storybook, we'll find this doesn't generalize to other situations quite as nicely, so I'll probably re-implement this at some point.

File: src/editor/reducers/file-source-inputs-reducer.js [hunk 27]
```diff
@@ -0,0 +1,32 @@
+function produceFileSourceInputs(state) {
```
Comment (hamilton) [inline src/editor/reducers/file-source-inputs-reducer.js]: it was requested to add a reducer specifically for the inputs for the `AddFileSource` component state. I am sure there will be further feedback on this PR that will necessitate more changes, so I will get to this before that next update or during it.

**Comment by hamilton** [comment]
@bcolloran as suggested, I removed all the last bits of local state from the file sources modal.

**Comment by hamilton** [comment]
I'd like to do a self-review of this PR this morning and then will ping again when I think it's ready for another pair of eyes.

File: docs/workflows.md [hunk 28]
```diff
@@ -27,13 +27,15 @@ If you want to work with data stored on your local machine or if you are
...
+### uploading files directly
+
 Data is uploaded on a per-notebook basis&mdash;rather than uploading data to
 your user account, you upload it to a specific notebook. To do so, click on the
 hamburger menu icon near the top-left of the page, then click _Manage Files_.
-The file management modal will appear. Click _Add Files_ and select any number
-of files to upload. Once the files finished uploading, they will appear in a
-list alongside other files associated with that notebook. Click _Delete_ at any
-time to delete a file.
+The file management modal will appear. By default, you will be in the _Manage
```
Comment (hamilton) [inline docs/workflows.md]: Could use some eyes on this docs blurb.

File: src/editor/components/modals/file-modal/manage-file-sources/file-source-list-item.jsx [hunk 29]
```diff
@@ -0,0 +1,245 @@
...
+  setConfirmDeleteID,
+  setIsDeletingAnimationID
+}) {
+  return (
+    <>
+      <DeleteModal
+        visible={isConfirmingDelete}
+        title={`Delete the file source for "${filename}"?`}
+        content="This action will not delete any files downloaded by the file source."
+        onCloseOrCancel={() => {}}
```
Comment (hamilton) [inline src/editor/components/modals/file-modal/manage-file-sources/file-source-list-item.jsx]: FIXME: this got removed while I was refactoring. We need an `onCloseOrCancel` function to be defined here before this merges. WIll do this today.

File: src/editor/components/modals/file-modal/manage-file-sources/validators.js [hunk 30]
```diff
@@ -0,0 +1,39 @@
+const ALLOWED_PROTOCOLS = ["https://", "http://"];
```
Comment (hamilton) [inline src/editor/components/modals/file-modal/manage-file-sources/validators.js]: I'm not sure if this file should go into some other shared dir, but I think for now it should be as closely colocated to the code that uses these functions.

**Comment by hamilton** [comment]
@bcolloran ok, at this point this should be ready for another review. We'll make sure to fix those merge conflicts before this lands but they are minor & shouldn't cause any friction anywhere.

File: server/files/api_views.py [hunk 31]
```diff
@@ -53,3 +61,79 @@ def update(self, request, pk):
...
+        return {"notebook_id": notebook_id}
+
+    def get_queryset(self):
+        base = FileSource.objects.filter(notebook_id=self.kwargs["notebook_id"])
+        filter_by_id = self.request.query_params.getlist("id")
+        if filter_by_id:
+            return base.filter(id__in=filter_by_id)
+        return base
+
+    queryset = FileSource.objects.all()
```
Comment (jezdez) [inline server/files/api_views.py]: Is `queryset` and `get_queryset` needed here?

File: server/files/models.py [hunk 32]
```diff
@@ -24,3 +26,63 @@ class Meta:
...
+    failure_reason = models.CharField(max_length=128, null=True)
+
+    def __str__(self):  # pragma: no cover
+        return "{} update ({})".format(self.file_source, self.OPERATION_STATUSES[self.status][1])
+
+    class Meta:
+        verbose_name = "File Update Operation"
+        verbose_name_plural = "File Update Operations"
+        ordering = ("id",)
+        db_table = "file_operation"
```
Comment (jezdez) [inline server/files/models.py]: shouldn't this be called `file_update_operation` to stay consistent?

File: server/files/models.py [hunk 33]
```diff
@@ -24,3 +26,63 @@ class Meta:
...
+    FAILED = 3
+    OPERATION_STATUSES = (
+        (PENDING, "pending"),
+        (RUNNING, "running"),
+        (COMPLETED, "completed"),
+        (FAILED, "failed"),
+    )
+
+    file_source = models.ForeignKey(FileSource, on_delete=models.CASCADE)
+    scheduled = models.DateTimeField(auto_now_add=True)
```
Comment (jezdez) [inline server/files/models.py]: It's a good practice to name the datetime fields `scheduled_at`, `started_at` and `ended_at` to indicate that type they are and are not confused with boolean fields.

File: server/files/models.py [hunk 34]
```diff
@@ -24,3 +26,63 @@ class Meta:
...
+    PENDING = 0
+    RUNNING = 1
+    COMPLETED = 2
+    FAILED = 3
+    OPERATION_STATUSES = (
+        (PENDING, "pending"),
+        (RUNNING, "running"),
+        (COMPLETED, "completed"),
+        (FAILED, "failed"),
+    )
```
Comment (jezdez) [inline server/files/models.py]: I don't think there is a huge problem with using string based values for the various operation statuses, and they would allow easier analysis later via Redash without having to handle those integer values. It's nicer to inspect in the database, too.

File: server/files/models.py (hunk 33 above)
Comment (jezdez) [reply server/files/models.py]: that makes sense, and granted, this was me just spelunking through the code, so no pressing need for an actual change here

Comment (wlach) [reply server/files/models.py]: I'll fix it, easier to do this now than later.
